File: 1-patch-match/patchmatch.py
#!/usr/bin/env python2

import numpy as np
import cv2
from visual import parser, load_frames, flow2rgb, show
import random
from itertools import product

import sys
def echo(*args, **kwargs):
    end = kwargs.get('end', '\n')

    if len(args) == 1:
        sys.stdout.write(args[0])
    else:
        for arg in args:
            sys.stdout.write(str(arg))
            sys.stdout.write(' ')

    sys.stdout.write(end)


def ssd(image1, center1, image2, center2, size):
    window1 = image1[center1[0] - size : center1[0] + size,
                     center1[1] - size : center1[1] + size]

    window2 = image2[center2[0] - size : center2[0] + size,
                     center2[1] - size : center2[1] + size]

    diff = window1 - window2

    return np.sum(diff ** 2)


SEARCH_FIELD = np.array(((-1, -1), (-1, 0), (-1, 1),
                         ( 0, -1),          ( 0, 1),
                         ( 1, -1), ( 1, 0), ( 1, 1)))


class PatchMatch(object):
    """Dump implementation of the PatchMatch algorithm as described by

    Connelly Barnes, Eli Shechtman, Adam Finkelstein, and Dan B. Goldman.
    PatchMatch: A randomized correspondence algorithm for structural image editing.
    In ACM Transactions on Graphics (Proc. SIGGRAPH), 2009. 2
    """

    def __init__(self, image1, image2, match_radius=5, search_ratio=0.5, search_radius=None, maxoffset=10):
        if not 0 < search_ratio < 1:
            raise ValueError('Search ratio must be in interval (0,1)')

        # input
        self.image1 = image1
        self.image2 = image2
        self.nrows, self.ncols = self.image1.shape

        self.niterations = 0

        # parameters
        self.maxoffset = maxoffset
        self.match_radius = match_radius
        self.search_ratio = search_ratio
        self.search_radius = search_radius or min(image1.shape)

        self.border = self.match_radius + self.maxoffset

        # create an empty matrix with the same x-y dimensions like the first
        # image but with two channels. Each channel stands for an x/y offset
        # of a pixel at this position.
        self.result  = np.zeros(dtype=np.int16, shape=(self.nrows, self.ncols, 2))
        self.quality = np.zeros(dtype=np.float32, shape=(self.nrows, self.ncols))

        # initialize offsets randomly
        self.initialize()

    def __iter__(self):
        rows = xrange(self.border, self.nrows - self.border)
        cols = xrange(self.border, self.ncols - self.border)

        for index in product(rows, cols):
            yield index

    def initialize(self):
        for index in self:
            # create a random offset in 
            offset = random.randint(-self.maxoffset, self.maxoffset), random.randint(-self.maxoffset, self.maxoffset)

            # assing random offset
            self.result[index] = offset

            # calculate the center in the second image by adding the offset
            # to the current index
            center = index[0] + offset[0], index[1] + offset[1]

            self.quality[index] = ssd(self.image1, index, self.image2, center, self.match_radius)

    def iterate(self):
        self.niterations += 1

        # switch between top and left neighbor in even iterations and
        # right bottom neighbor in odd iterations
        neighbor = -1 if self.niterations % 2 == 0 else 1

        for index in self:
            # echo('\r', end='')
            # echo('index', index, end='')
            self.propagate(index, neighbor)
        for index in self:
            self.random_search(index)

    def propagate(self, index, neighbor):
        indices = (index,                           # current position
                   (index[0] + neighbor, index[1]), # top / bottom neighbor
                   (index[0], index[1] + neighbor)) # left / right neighbor
    
        # create an array of all qualities at the above indices
        qualities = np.array((self.quality[indices[0]],
                              self.quality[indices[1]],
                              self.quality[indices[2]]))

        # get the index of the maximal quality
        maxindex = indices[np.argmax(qualities)]

        # get the offset of the neighbor with the maximal quality
        if maxindex != index:
            self.result[index] = self.result[maxindex]

    def random_search(self, index):
        i = 0
        quality = self.quality[index]

        while True:
            distance = int(self.search_radius * self.search_ratio ** i)
            i += 1

            # halt condition. search radius must not be smaller
            # than one pixel
            if distance < 1:
                break

            offset  = distance * random.choice(SEARCH_FIELD)
            center  = index + offset

            if self.border < center[0] < self.nrows - self.border and \
               self.border < center[1] < self.ncols - self.border:
                new_quality = ssd(self.image1, index, self.image2, center, self.match_radius)

                if new_quality > quality:
                    self.result[index] = offset
                    quality = new_quality


def reconstruct_from_flow(flow, image):
    result = np.zeros_like(image)

    for index in np.ndindex(flow.shape[0], flow.shape[1]):
        offset = flow[index]
        pixel  = index[0] + offset[0], index[1] + offset[1]
        result[index] = image[pixel]

    return result


if __name__ == '__main__':
    try:
        # command line parsing
        args = parser.parse_args()
        frame1, frame2 = load_frames(args.image1, args.image2)

        print('initialize ...')
        pm = PatchMatch(frame1, frame2)

        # result after initialziation
        show(reconstruct_from_flow(pm.result, frame2))
        show(flow2rgb(np.float32(pm.result)))

        # do some iterations
        for i in xrange(1):
            print('iteration %d ...' % (i + 1))
            pm.iterate()

            show(reconstruct_from_flow(pm.result, frame2))

            # display final result
            # we have to convert the integer offsets to floats, because
            # optical flow could be subpixel accurate
            rgb = flow2rgb(np.float32(pm.result))
            show(rgb)

    except KeyboardInterrupt:
        print('Stopping ...')

File: 1-patch-match/patchmatch.cpp
#include <opencv2/opencv.hpp>
#include <cstdlib>     // rand
#include <iostream>
#include <numeric> // numeric_limits
#include "patchmatch.hpp"

using namespace std;
using namespace cv;

float ssd(const Mat& image1, const Point2i& center1, const Mat& image2, const Point2i& center2, const int radius, const float halt)
{
    float sum = 0;

    for (int row = -radius; row <= radius; ++row) {
        for (int col = -radius; col <= radius; ++col) {
            // cout << (row + center1.y) << "," << (col + center1.x) << endl;
            // cout << (row + center2.y) << "," << (col + center2.x) << endl;
    
            const uchar gray1 = image1.at<uchar>(row + center1.y, col + center1.x);
            const uchar gray2 = image2.at<uchar>(row + center2.y, col + center2.x);

            const float diff  = gray1 - gray2;

            sum += diff * diff;

            // early termination
            if (sum > halt) {
                return sum;
            }

        }
    }

//    cout << sum << endl;

    return sum;
}


void flow2rgb(const Mat& flow, Mat& rgb)
{
    // extract x and y channels
    Mat xy[2]; // x,y

    split(flow, xy);

    // calculate angle and magnitude for the HSV color wheel
    Mat magnitude;
    Mat angle;

    cartToPolar(xy[0], xy[1], magnitude, angle, true);

    // translate magnitude to range [0,1]
    double mag_max;
    minMaxLoc(magnitude, 0, &mag_max);

    magnitude.convertTo(
        magnitude,    // output matrix
        -1,           // type of the ouput matrix, if negative same type as input matrix
        1.0 / mag_max // scaling factor
    );

    // build HSV image (hue-saturation-value)
    Mat _hsv[3]; // array of three matrices - one for each channel
    Mat hsv;

    // create separate channels
    _hsv[0] = angle;                           // H (hue)              [0,360]
    _hsv[1] = magnitude;                       // S (saturation)       [0,1]
    _hsv[2] = Mat::ones(angle.size(), CV_32F); // V (brigthness value) [0,1]

    // merge the three components to a three channel HSV image
    merge(_hsv, 3, hsv);

    // convert to BGR
    cvtColor(hsv, rgb, cv::COLOR_HSV2BGR);
}



PatchMatch::PatchMatch(int maxoffset, int match_radius, int iterations, float search_ratio, int search_radius) :
    // Parameters
    iterations(iterations),
    maxoffset(maxoffset),
    match_radius(match_radius),
    search_ratio(search_ratio),
    border(match_radius),
    max_search_radius(search_radius == -1)
{
    this->search_radius = search_radius;
}

void PatchMatch::match(const Mat& image1, const Mat& image2, Mat& dest)
{
    nrows = image1.rows;
    ncols = image1.cols;

    if (max_search_radius == true) {
        search_radius = min(nrows, ncols);
    }

    cout << "search_radius: " << search_radius << endl;

    // create an empty matrix with the same x-y dimensions like the first
    // image but with two channels. Each channel stands for an x/y offset
    // of a pixel at this position.
    costs = Mat::zeros(nrows, ncols, CV_32FC1);
    flow  = Mat::zeros(nrows, ncols, CV_32FC2); // 2-channel 32-bit floating point

    initialize(image1, image2);

    // for (niterations = 0; niterations < iterations; ++niterations) {

    //     cout << "iteration " << niterations << endl;

    //     for (int row = match_radius; row < nrows - match_radius ; ++row) {

    //         // cerr << row << endl;

    //         for (int col = match_radius; col < ncols - match_radius; ++col) {

    //             // cout << "    " << col << endl;

    //             propagate(image1, image2, row, col);
    //             // random_search(image1, image2, row, col);
    //         }
    //     }
    // }

    flow.copyTo(dest);
}

void PatchMatch::initialize(const Mat& image1, const Mat& image2)
{
    Point2i offset;
    Point2i index;
    Point2i center;

    for (int row = border; row < nrows - border ; ++row) {
        for (int col = border; col < ncols - border ; ++col) {
            index.x = col;
            index.y = row;

            // cerr << "index " << index << endl;

            while (true) {
                offset.x = rand() % (2 * maxoffset) - maxoffset;
                offset.y = rand() % (2 * maxoffset) - maxoffset;

                // cerr << "offset " << offset << endl;

                center = index + offset;

                // check if the center is inside the other image
                if (border <= center.x && center.x < ncols - border &&
                    border <= center.y && center.y < nrows - border
                ) {
                    // cerr << "break" << endl;
                    break;
                } else {
                    // cerr << index << " + " << offset << " = " << center << endl;

                    // cerr << border << " <= " << center.x << " && " << center.x << " < " << ncols << " - " << border << endl;
                    // cerr << border << " <= " << center.y << " && " << center.y << " < " << nrows << " - " << border << endl;
                }
            }

            // cerr << "offset " << offset << endl;

            flow.at<Point2f>(row, col) = offset;
            costs.at<float>(row, col) = ssd(image1, index, image2, center, match_radius);
        }
    }
}

void PatchMatch::propagate(const cv::Mat &image1, const cv::Mat &image2, const int row, const int col)
{
    // switch between top and left neighbor in even iterations and
    // right bottom neighbor in odd iterations
    int direction = (niterations % 2 == 0) ? 1 : -1;

    // top or bottom
    if (costs.at<float>(row + direction, col) < costs.at<float>(row, col)) {
        flow.at<Point2f>(row, col) = flow.at<Point2f>(row + direction, col);
    }
    // left or right
    if (costs.at<float>(row, col + direction) < costs.at<float>(row, col)) {
        flow.at<Point2f>(row, col) = flow.at<Point2f>(row, col + direction);
    }
}

static Point2f SEARCH_FIELD[8] = {
    Point2f(-1, -1), Point2f(-1, 0), Point2f(-1, 1),
    Point2f( 0, -1),                 Point2f( 0, 1),
    Point2f( 1, -1), Point2f( 1, 0), Point2f( 1, 1)
};

void PatchMatch::random_search(const cv::Mat &image1, const cv::Mat &image2, const int row, const int col)
{
    int i = 0;

    while (true) {
        const float distance = search_radius * pow(search_ratio, i++);

        // halt condition. search radius must not be smaller
        // than one pixel
        if (distance < 1) {
            break;
        }

        // cout << col << " " <<  distance << endl;

        const Point2f direction = SEARCH_FIELD[rand() % 8];
        const Point2f offset(direction.x * distance,
                             direction.y * distance);

        const Point2i center(row + (int) offset.x,
                             col + (int) offset.y);

        // check if we are inside image range
        if (match_radius < center.x && center.x < nrows - match_radius &&
            match_radius < center.y && center.y < ncols - match_radius) {

             float q = ssd(image1, Point2i(row, col), image2, center, search_radius, costs.at<float>(row, col));

//            float q = 0;

            // only update if the current match is better
//            if (q < quality.at<float>(row, col)) {
//                quality.at<float>(row, col) = q;
//                flow.at<Point2f>(row, col) = offset;
//            }
        }
    }
}
